src/react-components/link-root.js

```diff
@@ -0,0 +1,157 @@
+import React, { Component } from "react";
+import PropTypes from "prop-types";
+import { IntlProvider, FormattedMessage, addLocaleData } from "react-intl";
+import en from "react-intl/locale-data/en";
+
+import { lang, messages } from "../utils/i18n";
+import classNames from "classnames";
+import styles from "../assets/stylesheets/link.scss";
+
+const MAX_DIGITS = 4;
+
+addLocaleData([...en]);
+
+class LinkRoot extends Component {
+  static propTypes = {
+    intl: PropTypes.object,
+    store: PropTypes.object,
+    linkChannel: PropTypes.object
+  };
+
+  state = {
+    enteredDigits: [],
+    failedAtLeastOnce: false
+  };
+
+  componentWillMount = () => {
+    document.addEventListener("keydown", this.handleKeyDown);
+  };
+
+  componentWillUnmount = () => {
+    document.removeEventListener("keydown", this.handleKeyDown);
+  };
+
+  handleKeyDown = e => {
+    if (e.keyCode < 48 || e.keyCode > 57) {
+      return;
+    }
+
+    e.preventDefault();
+    e.stopPropagation();
+
+    this.addDigit(e.keyCode - 48);
+  };
+
+  addDigit = digit => {
+    if (this.state.enteredDigits.length >= MAX_DIGITS) return;
+    const newDigits = [...this.state.enteredDigits, digit];
+
+    if (newDigits.length === MAX_DIGITS) {
+      this.attemptLink(newDigits.join(""));
+    }
+
+    this.setState({ enteredDigits: newDigits });
+  };
+
+  removeDigit = () => {
+    if (this.state.enteredDigits.length === 0) return;
+    this.setState({ enteredDigits: [...this.state.enteredDigits.slice(0, -1)] });
+  };
+
+  attemptLink = code => {
+    this.props.linkChannel
+      .attemptLink(code)
+      .then(response => {
+        // If there is a profile from the linked device, copy it over if we don't have one yet.
+        if (response.profile) {
+          const { hasChangedName } = this.props.store.state.activity;
+
+          if (!hasChangedName) {
+            this.props.store.update({ activity: { hasChangedName: true }, profile: response.profile });
+          }
+        }
+
+        if (response.path) {
+          window.location.href = response.path;
+        }
+      })
+      .catch(e => {
+        console.error(e);
+        this.setState({ failedAtLeastOnce: true, enteredDigits: [] });
```

Comment:
Would we want sentry/metrics/measurements on failures?

Reply:
I didn't think so, since this is an expected condition if they mistype the code. There might be value in other cases, like multiple failures or something though.